### PR TITLE
Add a flutter doctor --android-licenses command that locates and runs the Android SDK license manager

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/8438

### CODE REVIEW

**Comment by jason-simmons** [comment]
@mit-mit @jakobr-google

File: packages/flutter_tools/lib/src/doctor.dart [hunk 1]
```diff
@@ -106,7 +106,11 @@ class Doctor {
   }
 
   /// Print verbose information about the state of installed tooling.
-  Future<bool> diagnose() async {
+  Future<bool> diagnose({bool androidLicenses: false}) async {
+    if (androidLicenses) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/doctor.dart]: nit: no braces

File: packages/flutter_tools/lib/src/doctor.dart [hunk 2]
```diff
@@ -106,7 +106,11 @@ class Doctor {
   }
 
   /// Print verbose information about the state of installed tooling.
-  Future<bool> diagnose() async {
+  Future<bool> diagnose({bool androidLicenses: false}) async {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/doctor.dart]: nitty nit nit: spaces around named arguments

**Comment by Hixie** [comment]
LGTM, but travis is angry.

I assume the plan is that a follow-up patch will replace the current large and confusing message from gradle with a much briefer message pointing to this?

**Comment by jason-simmons** [comment]
Fixed style issues and analytics_test

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 3]
```diff
@@ -163,4 +163,32 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
   }
+
+  /// Run the Android SDK manager tool in order to accept SDK licenses.
+  static Future<bool> runLicenseManager() async {
+    if (androidSdk == null) {
+      printStatus('Unable to locate Android SDK.');
+      return false;
+    }
+
+    // If we can locate Java, then add it to the path used to run the Android SDK manaer.
```
Comment (jakobr-google) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: s/manaer/manager/

**Review by jakobr-google — APPROVED** [review]
LGTM. Thanks!

**Comment by mit-mit** [comment]
I just upgraded on my clean test machine, and now `flutter doctor` is failing with:

```
## flutter doctor

encountered exception: NoSuchMethodError: The method 'contains' was called on null.
Receiver: null
Tried calling: contains("/")

#0      Object._noSuchMethod (dart:core-patch/object_patch.dart:43)
#1      Object.noSuchMethod (dart:core-patch/object_patch.dart:47)
#2      getExecutablePath (package:process/src/interface/common.dart:30)
#3      LocalProcessManager.canRun (package:process/src/interface/local_process_manager.dart:99)
#4      AndroidStudio._init (package:flutter_tools/src/android/android_studio.dart:276)
#5      AndroidStudio.AndroidStudio (package:flutter_tools/src/android/android_studio.dart:54)
#6      AndroidStudio.AndroidStudio.fromMacOSBundle (package:flutter_tools/src/android/android_studio.dart:74)
#7      AndroidStudio._allMacOS.<anonymous closure> (package:flutter_tools/src/android/android_studio.dart:184)
#8      MappedListIterable.elementAt (dart:_internal/iterable.dart:413)
#9      ListIterator.moveNext (dart:_internal/iterable.dart:342)
#10     WhereIterator.moveNext (dart:_internal/iterable.dart:437)
#11     List.List.from (dart:core-patch/array_patch.dart:46)
#12     Iterable.toList (dart:core/iterable.dart:347)
#13     AndroidStudio._allMacOS (package:flutter_tools/src/android/android_studio.dart:186)
#14     AndroidStudio.allInstalled (package:flutter_tools/src/android/android_studio.dart:143)
#15     AndroidStudio.latestValid (package:flutter_tools/src/android/android_studio.dart:133)
#16     AppContext.putIfAbsent (package:flutter_tools/src/base/context.dart:46)
#17     androidStudio (package:flutter_tools/src/android/android_studio.dart:17)
#18     javaPath (package:flutter_tools/src/android/android_studio.dart:49)
#19     AndroidWorkflow._findJavaBinary (package:flutter_tools/src/android/android_workflow.dart:37)
#20     AndroidWorkflow.validate (package:flutter_tools/src/android/android_workflow.dart:148)
<asynchronous suspension>
#21     Doctor.diagnose (package:flutter_tools/src/doctor.dart:116)
<asynchronous suspension>
#22     _doctorText.<anonymous closure> (package:flutter_tools/executable.dart:287)
#23     AppContext._run (package:flutter_tools/src/base/context.dart:76)
<asynchronous suspension>
#24     AppContext.runInZone.<anonymous closure> (package:flutter_tools/src/base/context.dart:66)
#25     _rootRun (dart:async/zone.dart:1120)
#26     _CustomZone.run (dart:async/zone.dart:1001)
#27     runZoned (dart:async/zone.dart:1465)
#28     AppContext.runInZone (package:flutter_tools/src/base/context.dart:65)
#29     _doctorText (package:flutter_tools/executable.dart:287)
<asynchronous suspension>
#30     _createLocalCrashReport (package:flutter_tools/executable.dart:262)
<asynchronous suspension>
#31     _handleToolError (package:flutter_tools/executable.dart:217)
<asynchronous suspension>
#32     run.<anonymous closure> (package:flutter_tools/executable.dart:144)
<asynchronous suspension>
#33     AppContext._run (package:flutter_tools/src/base/context.dart:76)
<asynchronous suspension>
#34     AppContext.runInZone.<anonymous closure> (package:flutter_tools/src/base/context.dart:66)
#35     _rootRun (dart:async/zone.dart:1120)
#36     _CustomZone.run (dart:async/zone.dart:1001)
#37     runZoned (dart:async/zone.dart:1465)
#38     AppContext.runInZone (package:flutter_tools/src/base/context.dart:65)
#39     run (package:flutter_tools/executable.dart:113)
<asynchronous suspension>
#40     main (package:flutter_tools/executable.dart:66)
<asynchronous suspension>
#41     main (file:///Users/myuser/flutter/packages/flutter_tools/bin/flutter_tools.dart:16)
#42     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#43     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)

```

@jason-simmons does that look related to this change?

**Comment by mit-mit** [comment]
Opened https://github.com/flutter/flutter/issues/9929

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 4]
```diff
@@ -163,4 +163,32 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
...
+
+    // If we can locate Java, then add it to the path used to run the Android SDK manager.
+    final Map<String, String> sdkManagerEnv = <String, String>{};
+    final String javaBinary = _findJavaBinary();
+    if (javaBinary != null) {
+      sdkManagerEnv['PATH'] =
+          platform.environment['PATH'] + os.pathVarSeparator + fs.path.dirname(javaBinary);
+    }
+
+    final Process process = await Process.start(
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: This should be using processManager, or more likely, methods in `base/process.dart`

**Pull request merged**